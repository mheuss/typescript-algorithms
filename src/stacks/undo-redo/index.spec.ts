import { UndoRedo } from "./index";

const UNDO_STACK = "undoStack";
const REDO_STACK = "redoStack";

describe("Undo/Redo Stack Tests", () => {
  describe("Place items into undo", () => {
    it("Should allow us to place items into the undo stack", () => {
      const undoStack = new UndoRedo<string>();
      undoStack.put("Drink Alcohol");
      expect(undoStack.undoCount()).toEqual(1);
      expect(undoStack.redoCount()).toEqual(0);
      expect(undoStack[UNDO_STACK].toArray()).toMatchSnapshot();
    });
  });

  describe("Undo nothing", () => {
    it("Should return undefined if there is nothing to undo", () => {
      const undoStack = new UndoRedo<string>();
      expect(undoStack.undo()).toBeUndefined();
    });
  });

  describe("Redo nothing", () => {
    it("Should return undefined if there is nothing to Redo", () => {
      const undoStack = new UndoRedo<string>();
      expect(undoStack.redo()).toBeUndefined();
    });
  });

  describe("Undo Items in sequence", () => {
    const undoStack = new UndoRedo<string>();

    it("Should allow us to place items into the undo stack", () => {
      undoStack.put("Drink Alcohol");
      undoStack.put("Sing Loudly");
      undoStack.put("Dance With Abandon");
      undoStack.put("Drink More");
      undoStack.put("Drive Towards Home");
      undoStack.put("Run into tree");
      expect(undoStack.undoCount()).toEqual(6);
      expect(undoStack.redoCount()).toEqual(0);
      expect(undoStack[UNDO_STACK].toArray()).toMatchSnapshot();
    });

    it("By undoing, we should be populating the redo stack", () => {
      expect(undoStack.undo()).toEqual("Run into tree");
      expect(undoStack.undo()).toEqual("Drive Towards Home");
      expect(undoStack.undo()).toEqual("Drink More");
      expect(undoStack.undo()).toEqual("Dance With Abandon");
      expect(undoStack.undo()).toEqual("Sing Loudly");
      expect(undoStack.undo()).toEqual("Drink Alcohol");
      expect(undoStack.undoCount()).toEqual(0);
      expect(undoStack.redoCount()).toEqual(6);
      expect(undoStack[REDO_STACK].toArray()).toMatchSnapshot();
    });

    it("By redoing, we should be populating the undo stack", () => {
      expect(undoStack.redo()).toEqual("Drink Alcohol");
      expect(undoStack.redo()).toEqual("Sing Loudly");
      expect(undoStack.redo()).toEqual("Dance With Abandon");
      expect(undoStack.undoCount()).toEqual(3);
      expect(undoStack.redoCount()).toEqual(3);
      expect(undoStack[REDO_STACK].toArray()).toMatchSnapshot();
    });
  });

  describe("Test of redo clear when new item pushed into undo", () => {
    const undoStack = new UndoRedo<string>();
    it("Let us get stack parity", () => {
      undoStack.put("Drink Alcohol");
      undoStack.put("Sing Loudly");
      undoStack.put("Dance With Abandon");
      undoStack.put("Drink More");
      undoStack.put("Drive Towards Home");
      undoStack.put("Run into tree");
      undoStack.undo();
      undoStack.undo();
      undoStack.undo();
      expect(undoStack.undoCount()).toEqual(3);
      expect(undoStack.redoCount()).toEqual(3);
    });

    it("Should clear redo when new item pushed", () => {
      undoStack.put("Pass Out");
      expect(undoStack.undoCount()).toEqual(4);
      expect(undoStack.redoCount()).toEqual(0);
    });
  });

  describe("Emptying stack", () => {
    const undoStack = new UndoRedo<string>();

    it("Should allow us to place items into the undo stack", () => {
      undoStack.put("Drink Alcohol");
      undoStack.put("Sing Loudly");
      undoStack.put("Dance With Abandon");
      undoStack.put("Drink More");
      undoStack.put("Drive Towards Home");
      undoStack.put("Run into tree");
      undoStack.undo();
      undoStack.undo();
      undoStack.undo();
      expect(undoStack.undoCount()).toEqual(3);
      expect(undoStack.redoCount()).toEqual(3);
    });

    it("And let us clear", () => {
      undoStack.clear();
      expect(undoStack.undoCount()).toEqual(0);
      expect(undoStack.redoCount()).toEqual(0);
    });
  });
});
