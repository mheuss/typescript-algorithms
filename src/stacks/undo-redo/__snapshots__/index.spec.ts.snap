// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Undo/Redo Stack Tests Place items into undo Should allow us to place items into the undo stack 1`] = `
Array [
  "Drink Alcohol",
]
`;

exports[`Undo/Redo Stack Tests Undo Items in sequence By redoing, we should be populating the undo stack 1`] = `
Array [
  "Drink More",
  "Drive Towards Home",
  "Run into tree",
]
`;

exports[`Undo/Redo Stack Tests Undo Items in sequence By undoing, we should be populating the redo stack 1`] = `
Array [
  "Drink Alcohol",
  "Sing Loudly",
  "Dance With Abandon",
  "Drink More",
  "Drive Towards Home",
  "Run into tree",
]
`;

exports[`Undo/Redo Stack Tests Undo Items in sequence Should allow us to place items into the undo stack 1`] = `
Array [
  "Run into tree",
  "Drive Towards Home",
  "Drink More",
  "Dance With Abandon",
  "Sing Loudly",
  "Drink Alcohol",
]
`;
